### PR TITLE
inotify-util: drop to use pointer outside of the buffer

### PR DESCRIPTION
**Pull request by yuwata**
Fixes many Coverity issues.

### CODE REVIEW

File: src/basic/inotify-util.h [hunk 1]
```diff
@@ -10,29 +10,58 @@
...
 #define FOREACH_INOTIFY_EVENT_WARN(e, buffer, sz)               \
-        _FOREACH_INOTIFY_EVENT_FULL(e, buffer, sz, LOG_WARNING)
+        FOREACH_INOTIFY_EVENT_FULL(e, buffer, sz, LOG_WARNING)
 
 union inotify_event_buffer {
         struct inotify_event ev;
         uint8_t raw[INOTIFY_EVENT_MAX];
 };
 
+static inline bool inotify_event_next(
```
Comment (bluca) [inline src/basic/inotify-util.h]: probably a bit too long and complex for a static inline?

Comment (poettering) [reply src/basic/inotify-util.h]: yeah, i'd agree, this is a big too much for an inline helper

File: src/basic/inotify-util.h [hunk 2]
```diff
@@ -10,29 +10,58 @@
...
+        assert(buffer);
+        assert(iterator);
+
+        if (*iterator) {
+                assert((uint8_t*) *iterator >= buffer->raw);
+                offset = (uint8_t*) *iterator - buffer->raw;
+                offset += offsetof(struct inotify_event, name) + (*iterator)->len;
+        }
+
+        if (size == offset)
```
Comment (poettering) [inline src/basic/inotify-util.h]: maybe add a comment `/* reached end of list */`

File: src/basic/inotify-util.h [hunk 3]
```diff
@@ -10,29 +10,58 @@
...
+                return false;
+
+        if (size < offset ||
+            size - offset < offsetof(struct inotify_event, name)) {
+                log_full(log_level, "Received invalid inotify event, ignoring.");
+                return false;
+        }
+
+        e = (struct inotify_event*) (buffer->raw + offset);
```
Comment (poettering) [inline src/basic/inotify-util.h]: maybe use CAST_ALIGN_PTR() for extra safety?

**Comment by poettering** [comment]
lgtm

**Comment by yuwata** [comment]
@bluca @poettering Thank you for review. All suggestions are applied. Upgrading the green label.

**Comment by keszybz** [comment]
jammy-amd64: 8501s TEST-29-PORTABLE:                   FAIL     (179 s), probably unrelated.

**Pull request merged**